### PR TITLE
Add support for constructing large constant arrays

### LINKED ISSUE #4008: Array constructor with too many values causes ClassFormatError
Queries that contain very large array constructors cause bytecode to be generated that contains too many method arguments.

```
SELECT ARRAY['1','1','1','2','1','1','-1','1','1','1','1','1','-1','1','1','1','1','1','1','1','1','1','-1','1','-1','1','1','1','1','1','1','1','-1','-1','-1','1','-1','1','-1','3','-1','1','1','1','1','-1','-1','1','1','-1','-1','1','1','1','1','1','-1','1','1','1','1','1','-1','-1','1','-1','1','-1','1','1','1','-1','1','1','-1','1','1','-1','1','-1','-1','-1','1','1','-1','1','1','1','1','1','-1','1','-1','1','-1','1','1','1','1','-1','-1','-1','1','-1','-1','-1','1','1','1','-1','1','-1','-1','-1','-1','1','-1','-1','-1','1','-1','1','1','3','1','-1','-1','1','-1','-1','1','-1','-1','1','1','1','-1','-1','-1','1','1','1','-1','1','1','1','1','-1','-1','-1','1','-1','1','-1','3','-1','-1','1','-1','1','1','-1','-1','3','-1','1','-1','-1','-1','1','-1','-1','1','1','1','1','1','-1','1','1','-1','-1','-1','1','1','1','-1','1','-1','1','1','1','-1','-1','1','1','-1','-1','-1','-1','1','-1','-1','1','-1','1','1','-1','1','-1','-1','1','1','-1','-1','-1','-1','1','1','1','-1','1','1','1','1','1','1','1','1','1','1','1','1','1','1','-1','-1','-1','-1','1','-1','1','-1','-1','1','3','-1','-1','1','1','3','1','1','1','1','1','1','1','1','-1','1','1','3','3','1','1','1','-1','-1','1','1','1','3','1','1','1','1','1','1','1','1','1','1','1','1','1','1','-1','1','1','1','1','1','1','1','3','1','1','1','1','1','1','1','1','1','1','1','1','-1','1','1','1','1','1','-1','1','3','1','1','1','1','1','3','1','3','1','1','1','1','3','-1','-1','-1','1','1','1','-1','-1','-1','-1','-1','1','-1','1','1','3','1','1','1','1','-1','-1','3','1','1','3','-1','1','1','1','-1','-1','1','1','1','-1','1','1','1','-1','1','3','1','1','1','1','3','3','-1','1','1','-1','3','3','1','1','3','-1','-1','1','1','-1','-1','1','1','3','1','-1','3','3','1','1','3','1','1','1','3','3','1','1','3','3','3','1','1','1','1','-1','-1','1','1','3','1','1','3','3','1','1','3','3','1','-1','1','1','1','1','-1','1','1','1','1','1','3','3','-1','-1','-1','1','-1','1','1','3','1','3','1','1','1','1','1','3','1','1','-1','1','3','3','3','1','1','3','1','1','1','1','1','1','1','1','1','1','1','1','1','1','1','1','1','1','1','1','1','3','1','3','3','1','1','1','1','1','1','1','1','1','1','1','1','1','3','3','1','1'];
```
fails with
```
com.google.common.util.concurrent.ExecutionError: java.lang.ClassFormatError: Too many arguments in method signature i
n class file com_facebook_presto_$gen_SliceSliceSliceSliceSliceSliceSliceSliceSliceSliceSliceSliceSliceSliceSliceSlice
SliceSliceSliceSliceSliceSliceSliceSliceSliceSliceSliceSliceSliceSliceSliceSliceSliceSliceSliceSliceSliceSliceSliceSli
ceSliceSliceSliceSliceSliceSliceSliceSliceSliceSliceSliceSliceSliceSliceSliceSliceSliceSliceSliceSliceSliceSliceSliceS
liceSliceSliceSliceSliceSliceSliceSliceSliceSliceSliceSliceSliceSliceSliceSliceSliceSliceSliceSliceSliceSliceSliceSlic
eSliceSliceSliceSliceSliceSliceSliceSliceSliceSliceSliceSliceSliceSliceSliceSliceSliceSliceSliceSliceSliceSliceSliceSl
iceSliceSliceSliceSliceSliceSliceSliceSliceSliceSliceSliceSliceSliceSliceSliceSliceSliceSliceSliceSliceSliceSliceSlice
SliceSliceSliceSliceSliceSliceSliceSliceSliceSliceSliceSliceSliceSliceSliceSliceSliceSliceSliceSliceSliceSliceSliceSli
ceSliceSliceSliceSliceSliceSliceSliceSliceSliceSliceSliceSliceSliceSliceSliceSliceSliceSliceSliceSliceSliceSliceSliceS
liceSliceSliceSliceSliceSliceSliceSlice
        at com.google.common.cache.LocalCache$Segment.get(LocalCache.java:2201)
        at com.google.common.cache.LocalCache.get(LocalCache.java:3937)
        at com.google.common.cache.LocalCache.getOrLoad(LocalCache.java:3941)
        at com.google.common.cache.LocalCache$LocalLoadingCache.get(LocalCache.java:4824)
        at com.google.common.cache.LocalCache$LocalLoadingCache.getUnchecked(LocalCache.java:4830)
        at com.facebook.presto.metadata.FunctionRegistry.getScalarFunctionImplementation(FunctionRegistry.java:586)
        at com.facebook.presto.sql.planner.ExpressionInterpreter$Visitor.visitFunctionCall(ExpressionInterpreter.java:
798)
        at com.facebook.presto.sql.planner.ExpressionInterpreter$Visitor.visitArrayConstructor(ExpressionInterpreter.j
ava:940)
        at com.facebook.presto.sql.tree.ArrayConstructor.accept(ArrayConstructor.java:54)
        at com.facebook.presto.sql.tree.AstVisitor.process(AstVisitor.java:22)
        at com.facebook.presto.sql.planner.ExpressionInterpreter.optimize(ExpressionInterpreter.java:251)
        at com.facebook.presto.sql.planner.optimizations.SimplifyExpressions$Rewriter.simplifyExpression(SimplifyExpre
ssions.java:133)
        at com.google.common.collect.Maps$7.transformEntry(Maps.java:1812)
        at com.google.common.collect.Maps$10.getValue(Maps.java:1857)
        at com.google.common.collect.ImmutableMap.copyOf(ImmutableMap.java:292)
        at com.facebook.presto.sql.planner.optimizations.SimplifyExpressions$Rewriter.visitProject(SimplifyExpressions
.java:92)
        at com.facebook.presto.sql.planner.optimizations.SimplifyExpressions$Rewriter.visitProject(SimplifyExpressions
.java:70)
        at com.facebook.presto.sql.planner.plan.ProjectNode.accept(ProjectNode.java:81)
        at com.facebook.presto.sql.planner.plan.SimplePlanRewriter$RewriteContext.rewrite(SimplePlanRewriter.java:86)
        at com.facebook.presto.sql.planner.plan.SimplePlanRewriter$RewriteContext.rewrite(SimplePlanRewriter.java:97)
        at com.facebook.presto.sql.planner.optimizations.SimplifyExpressions$Rewriter.visitProject(SimplifyExpressions
.java:91)
        at com.facebook.presto.sql.planner.optimizations.SimplifyExpressions$Rewriter.visitProject(SimplifyExpressions
.java:70)
        at com.facebook.presto.sql.planner.plan.ProjectNode.accept(ProjectNode.java:81)
        at com.facebook.presto.sql.planner.plan.SimplePlanRewriter$RewriteContext.rewrite(SimplePlanRewriter.java:86)
        at com.facebook.presto.sql.planner.plan.SimplePlanRewriter$RewriteContext.rewrite(SimplePlanRewriter.java:97)
        at com.facebook.presto.sql.planner.optimizations.SimplifyExpressions$Rewriter.visitProject(SimplifyExpressions
.java:91)
...
```

### PR DESCRIPTION
**Pull request by raghavsethi**
Generating bytecode to constant fold array constructors fails when the number of elements in the array exceeds 255 (i.e. the maximum number of arguments to a Java function). We now simply generate an array block instead of generating bytecode.

Fixes #4008.

### CODE REVIEW

**Comment by haozhun** [comment]
checkstyle failure.

File: presto-main/src/main/java/com/facebook/presto/sql/planner/ExpressionInterpreter.java [hunk 1]
```diff
@@ -793,6 +799,7 @@ protected Object visitFunctionCall(FunctionCall node, Object context)
                 argumentValues.add(value);
                 argumentTypes.add(type);
             }
+
```
Comment (haozhun) [inline presto-main/src/main/java/com/facebook/presto/sql/planner/ExpressionInterpreter.java]: revert unrelated change

**Comment by raghavsethi** [comment]
Sorry, they're all rebase conflicts. Fixed.

File: presto-main/src/main/java/com/facebook/presto/sql/planner/ExpressionInterpreter.java [hunk 2]
```diff
@@ -936,7 +941,33 @@ public Object visitCast(Cast node, Object context)
         @Override
         protected Object visitArrayConstructor(ArrayConstructor node, Object context)
         {
-            return visitFunctionCall(new FunctionCall(QualifiedName.of(ArrayConstructor.ARRAY_CONSTRUCTOR), node.getValues()), context);
+            ImmutableSet.Builder<Type> elementTypes = ImmutableSet.builder();
```
Comment (haozhun) [inline presto-main/src/main/java/com/facebook/presto/sql/planner/ExpressionInterpreter.java]: `Type elementType = ((ArrayType) expressionTypes.get(node)).getElementType()`

File: presto-main/src/main/java/com/facebook/presto/sql/planner/ExpressionInterpreter.java [hunk 3]
```diff
@@ -936,7 +941,33 @@ public Object visitCast(Cast node, Object context)
...
+                if (value instanceof Expression) {
+                    containsExpression = true;
+                    break;
+                }
+                Type type = expressionTypes.get(expression);
+                arrayValues.add(value);
+                elementTypes.add(type);
+            }
+
+            if (containsExpression || arrayValues.isEmpty()) {
```
Comment (haozhun) [inline presto-main/src/main/java/com/facebook/presto/sql/planner/ExpressionInterpreter.java]: Now that you can tell the `elementType` even with empty array constructor, you don't need the second parameter any more.

File: presto-main/src/test/java/com/facebook/presto/sql/TestExpressionInterpreter.java [hunk 4]
```diff
@@ -937,6 +940,19 @@ public void testArraySubscriptConstantZeroIndex()
...
+    public void testMassiveArrayConstructorConstants()
+    {
+        optimize(format("ARRAY [%s]", Joiner.on(", ").join(IntStream.range(0, 10_000).mapToObj(i -> "(bound_long + " + i + ")").iterator())));
+    }
+
+    @Test
+    public void testArrayConstructorExpressions()
+    {
+        assertOptimizedEquals("ARRAY [(unbound_long + 0), (unbound_long + 1), (unbound_long + 2)]",
+                "array_constructor((unbound_long + 0), (unbound_long + 1), (unbound_long + 2))");
```
Comment (haozhun) [inline presto-main/src/test/java/com/facebook/presto/sql/TestExpressionInterpreter.java]: Add test for mixed case

File: presto-main/src/test/java/com/facebook/presto/sql/TestExpressionInterpreter.java [hunk 5]
```diff
@@ -937,6 +940,19 @@ public void testArraySubscriptConstantZeroIndex()
     }
 
     @Test
+    public void testMassiveArrayConstructorConstants()
+    {
+        optimize(format("ARRAY [%s]", Joiner.on(", ").join(IntStream.range(0, 10_000).mapToObj(i -> "(bound_long + " + i + ")").iterator())));
```
Comment (haozhun) [inline presto-main/src/test/java/com/facebook/presto/sql/TestExpressionInterpreter.java]: Add a test for variable-width type and a test for complex type.

File: presto-main/src/test/java/com/facebook/presto/sql/TestExpressionInterpreter.java (hunk 4 above)
Comment (raghavsethi) [inline presto-main/src/test/java/com/facebook/presto/sql/TestExpressionInterpreter.java]: add test for empty

File: presto-main/src/test/java/com/facebook/presto/sql/TestExpressionInterpreter.java (hunk 5 above)
Comment (raghavsethi) [inline presto-main/src/test/java/com/facebook/presto/sql/TestExpressionInterpreter.java]: Add test where null exists

**Comment by raghavsethi** [comment]
Back to you @haozhun

**Comment by haozhun** [comment]
We don't normally write one test method per test case. When there isn't a particular reason, let's follow the conventions.

I would merge the tests that involve massive into one method, and then tests that don't involve massive into another method.

**Comment by haozhun** [comment]
Looks good. Ask @erichwang if he wants to take a 2nd look.

File: presto-main/src/main/java/com/facebook/presto/sql/planner/ExpressionInterpreter.java [hunk 6]
```diff
@@ -936,7 +940,28 @@ public Object visitCast(Cast node, Object context)
         @Override
         protected Object visitArrayConstructor(ArrayConstructor node, Object context)
         {
-            return visitFunctionCall(new FunctionCall(QualifiedName.of(ArrayConstructor.ARRAY_CONSTRUCTOR), node.getValues()), context);
+            Type elementType = ((ArrayType) expressionTypes.get(node)).getElementType();
+            List<Object> arrayValues = new ArrayList<>();
```
Comment (erichwang) [inline presto-main/src/main/java/com/facebook/presto/sql/planner/ExpressionInterpreter.java]: maybe initialize the size of the array list with node.getValues().size()?

File: presto-main/src/main/java/com/facebook/presto/sql/planner/ExpressionInterpreter.java [hunk 7]
```diff
@@ -936,7 +940,28 @@ public Object visitCast(Cast node, Object context)
...
+                Object value = process(expression, context);
+                if (value instanceof Expression) {
+                    containsExpression = true;
+                    break;
+                }
+                arrayValues.add(value);
+            }
+
+            if (containsExpression) {
+                return visitFunctionCall(new FunctionCall(QualifiedName.of(ArrayConstructor.ARRAY_CONSTRUCTOR), node.getValues()), context);
```
Comment (erichwang) [inline presto-main/src/main/java/com/facebook/presto/sql/planner/ExpressionInterpreter.java]: i think you can just inline this return statement where you put the break statement. then you don't need the extra boolean.

**Comment by erichwang** [comment]
looks good otherwise
